contracts/PoolManager.sol

```diff
@@ -132,55 +132,69 @@ contract PoolManager is IPoolManager, Owned, NoDelegateCall, ERC1155, IERC1155Re
     /// @member nonzeroDeltaCount The number of entries in the currencyDelta mapping that have a non-zero value
     /// @member currencyDelta The amount owed to the locker (positive) or owed to the pool (negative) of the currency
     struct LockState {
         uint256 nonzeroDeltaCount;
         mapping(Currency currency => int256) currencyDelta;
     }
 
-    /// @dev Represents the state of the locker at the given index. Each locker must have net 0 currencies owed before
-    /// releasing their lock. Note this is private because the nested mappings cannot be exposed as a public variable.
-    mapping(uint256 index => LockState) private lockStates;
+    /// @dev Represents the state of the given locker. Each locker must have net 0 currencies due/owed after the
+    /// last lock is released. Note this is private because the nested mappings cannot be exposed as a public variable.
+    mapping(address locker => LockState) private lockStates;
 
     /// @inheritdoc IPoolManager
-    function getNonzeroDeltaCount(uint256 id) external view returns (uint256) {
-        return lockStates[id].nonzeroDeltaCount;
+    function getNonzeroDeltaCount(address locker) external view returns (uint256) {
+        return lockStates[locker].nonzeroDeltaCount;
     }
 
     /// @inheritdoc IPoolManager
-    function getCurrencyDelta(uint256 id, Currency currency) external view returns (int256) {
-        return lockStates[id].currencyDelta[currency];
+    function getCurrencyDelta(address locker, Currency currency) external view returns (int256) {
+        return lockStates[locker].currencyDelta[currency];
     }
 
     /// @inheritdoc IPoolManager
     function lock(bytes calldata data) external override returns (bytes memory result) {
-        uint256 id = lockedBy.length;
-        lockedBy.push(msg.sender);
-
-        // the caller does everything in this callback, including paying what they owe via calls to settle
-        result = ILockCallback(msg.sender).lockAcquired(id, data);
-
         unchecked {
-            LockState storage lockState = lockStates[id];
-            if (lockState.nonzeroDeltaCount != 0) revert CurrencyNotSettled();
+            locks.push(Lock({locker: msg.sender, parentLockIndex: uint96(lockIndex)}));
+            lockIndex = locks.length - 1;
+
+            // the caller does everything in this callback, including paying what they owe via calls to settle
+            result = ILockCallback(msg.sender).lockAcquired(data);
+
+            // only enforce that deltas are 0 when the outermost lock frame is expiring
+            if (lockIndex == 0) {
+                uint256 length = locksLength();
+                for (uint256 i; i < length; i++) {
+                    address locker = locks[i].locker;
+                    if (lockStates[locker].nonzeroDeltaCount != 0) revert CurrencyNotSettled(locker);
+                }
+                delete locks;
+            } else {
+                lockIndex = locks[lockIndex].parentLockIndex;
```

Comment:
what's the point of `parentLockIndex`?? is it ever not the previous Lock in the array?  🤔

Reply:
yeah so this was the problem with the previous version of this - the parentLockIndex is actually (non-intuitively) not always the previous lock. see `assertionChecker1` for an example (the parentLockIndex is 0, then 2 for the subsequent lock; this is because there was an opened + closed lock before it which bumps the index)